core/src/test/java/kafka/server/share/SharePartitionTest.java

```diff
@@ -490,11 +490,14 @@ public void testMaybeInitializeSharePartitionAgainConcurrentRequests() throws In
             if (!executorService.awaitTermination(30, TimeUnit.MILLISECONDS))
                 executorService.shutdown();
         }
+        int futuresCompleted = 0;
 
         for (CompletableFuture<Void> result : results) {
             assertTrue(result.isDone());
-            assertFalse(result.isCompletedExceptionally());
+            if (!result.isCompletedExceptionally())
```

Comment:
Pardon me, this PR's primary purpose is to add the write lock to maybeInitialize. Why are these additional changes needed?

Reply:
The test appears to assume that all threads will successfully initialize the share partition concurrently. However, due to the narrowed lock scope introduced in this PR, some threads are now failing because the partition is in the `INITIALIZING` state.

Reply:
Maybe we can simplify the code by lambda?
```
        assertTrue(results.stream().allMatch(CompletableFuture::isDone));
        assertFalse(results.stream().allMatch(CompletableFuture::isCompletedExceptionally));
```

Reply:
Hi @chia7712 , yes, if not for the change, this test would become highly flaky.

> Maybe we can simplify the code by lambda?
> 
> ```
>         assertTrue(results.stream().allMatch(CompletableFuture::isDone));
>         assertFalse(results.stream().allMatch(CompletableFuture::isCompletedExceptionally));
> ```
Made this change. Please review the PR. Thanks!